File: medieval_recolor_clean.js
const fs = require('fs');
const path = require('path');
const zlib = require('zlib');

function decodePNG(srcPath) {
    const data = fs.readFileSync(srcPath);
    let offset = 8;
    let width = 0, height = 0;
    let idatBuffers = [];

    while (offset < data.length) {
        const length = data.readUInt32BE(offset);
        const type = data.toString('ascii', offset + 4, offset + 8);
        const chunkData = data.slice(offset + 8, offset + 8 + length);

        if (type === 'IHDR') {
            width = chunkData.readUInt32BE(0);
            height = chunkData.readUInt32BE(4);
        } else if (type === 'IDAT') {
            idatBuffers.push(chunkData);
        }
        offset += 12 + length;
    }

    const raw = zlib.inflateSync(Buffer.concat(idatBuffers));
    const bpp = 4;
    const stride = width * bpp;
    const pixels = Buffer.alloc(height * stride);

    let rawOffset = 0;
    for (let y = 0; y < height; y++) {
        const filterType = raw[rawOffset++];
        const scanline = raw.slice(rawOffset, rawOffset + stride);
        rawOffset += stride;

        const destOffset = y * stride;
        for (let x = 0; x < stride; x++) {
            const filt = scanline[x];
            const left = x >= bpp ? pixels[destOffset + x - bpp] : 0;
            const up = y > 0 ? pixels[destOffset - stride + x] : 0;
            const upleft = (y > 0 && x >= bpp) ? pixels[destOffset - stride + x - bpp] : 0;

            let val = filt;
            if (filterType === 1) val = (filt + left) & 0xff;
            else if (filterType === 2) val = (filt + up) & 0xff;
            else if (filterType === 3) val = (filt + Math.floor((left + up) / 2)) & 0xff;
            else if (filterType === 4) {
                const p = left + up - upleft;
                const pa = Math.abs(p - left);
                const pb = Math.abs(p - up);
                const pc = Math.abs(p - upleft);
                let pr = left;
                if (pa <= pb && pa <= pc) pr = left;
                else if (pb <= pc) pr = up;
                else pr = upleft;
                val = (filt + pr) & 0xff;
            }
            pixels[destOffset + x] = val;
        }
    }
    return { width, height, pixels };
}

function encodePNG(width, height, pixels, dstPaths) {
    const stride = width * 4;
    const rawData = [];
    for (let y = 0; y < height; y++) {
        rawData.push(0); // filter 0
        for (let x = 0; x < stride; x++) {
            rawData.push(pixels[y * stride + x]);
        }
    }
    const idatData = zlib.deflateSync(Buffer.from(rawData));

    function createChunk(type, chunkData) {
        const len = Buffer.alloc(4);
        len.writeUInt32BE(chunkData.length, 0);
        const typeBuf = Buffer.from(type, 'ascii');
        const body = Buffer.concat([typeBuf, chunkData]);

        let crc = 0xFFFFFFFF;
        for (let i = 0; i < body.length; i++) {
            crc ^= body[i];
            for (let j = 0; j < 8; j++) {
                crc = (crc >>> 1) ^ (crc & 1 ? 0xEDB88320 : 0);
            }
        }
        crc = (crc ^ 0xFFFFFFFF) >>> 0;

        const crcBuf = Buffer.alloc(4);
        crcBuf.writeUInt32BE(crc, 0);
        return Buffer.concat([len, body, crcBuf]);
    }

    const signature = Buffer.from([0x89, 0x50, 0x4E, 0x47, 0x0D, 0x0A, 0x1A, 0x0A]);
    const ihdrBuf = Buffer.alloc(13);
    ihdrBuf.writeUInt32BE(width, 0);
    ihdrBuf.writeUInt32BE(height, 4);
    ihdrBuf[8] = 8; ihdrBuf[9] = 6; ihdrBuf[10] = 0; ihdrBuf[11] = 0; ihdrBuf[12] = 0;

    const newPNG = Buffer.concat([
        signature,
        createChunk('IHDR', ihdrBuf),
        createChunk('IDAT', idatData),
        createChunk('IEND', Buffer.alloc(0))
    ]);

    for (const dstPath of dstPaths) {
        fs.mkdirSync(path.dirname(dstPath), { recursive: true });
        fs.writeFileSync(dstPath, newPNG);
        console.log(`[OK] Created Clean Semi-Medieval PNG: ${dstPath}`);
    }
}

function processRecolor() {
    const srcPng = path.join(__dirname, 'kamish-dagger-1-20-1-e1560', 'assets', 'minecraft', 'optifine', 'cit', 'kamish_dagger', 'kamish_dagger.png');
    const itemDir = path.join(__dirname, 'src', 'main', 'resources', 'assets', 'corazonmod', 'textures', 'item');

    const { width, height, pixels } = decodePNG(srcPng);

    for (let i = 0; i < pixels.length; i += 4) {
        const r = pixels[i];
        const g = pixels[i+1];
        const b = pixels[i+2];
        const a = pixels[i+3];

        if (a < 10) continue;

        const lum = (0.299 * r + 0.587 * g + 0.114 * b) / 255.0;

        const isRedPurple = (r > 120 && b > 100) || (r > 150 && g < 100);
        const isYellowGold = (r > 140 && g > 120 && b < 100);

        let nr, ng, nb;

        if (isRedPurple) {
            if (lum > 0.6) {
                nr = Math.min(255, Math.floor(255 * lum));
                ng = Math.min(255, Math.floor(215 * lum));
                nb = Math.min(255, Math.floor(90 * lum));
            } else {
                nr = Math.min(255, Math.floor(200 * lum + 30));
                ng = Math.min(255, Math.floor(40 * lum));
                nb = Math.min(255, Math.floor(50 * lum));
            }
        } else if (isYellowGold) {
            nr = Math.min(255, Math.floor(lum * 230 + 30));
            ng = Math.min(255, Math.floor(lum * 170 + 20));
            nb = Math.min(255, Math.floor(lum * 60 + 10));
        } else {
            if (lum > 0.7) {
                nr = Math.min(255, Math.floor(lum * 240 + 15));
                ng = Math.min(255, Math.floor(lum * 245 + 10));
                nb = Math.min(255, Math.floor(lum * 255));
            } else if (lum > 0.35) {
                nr = Math.floor(lum * 170 + 20);
                ng = Math.floor(lum * 180 + 20);
                nb = Math.floor(lum * 200 + 25);
            } else {
                nr = Math.floor(lum * 100 + 15);
                ng = Math.floor(lum * 85 + 10);
                nb = Math.floor(lum * 75 + 10);
            }
        }

        pixels[i] = nr;
        pixels[i+1] = ng;
        pixels[i+2] = nb;
    }

    encodePNG(width, height, pixels, [
        path.join(itemDir, 'mafia_dagger.png'),
        path.join(itemDir, 'kamish_dagger.png')
    ]);
}

processRecolor();

File: medieval_recolor_clean.py
import struct
import zlib
import os

def decode_png(filepath):
    with open(filepath, 'rb') as f:
        data = f.read()
    
    assert data[:8] == b'\x89PNG\r\n\x1a\n'
    
    offset = 8
    width = 0
    height = 0
    idat_chunks = []
    
    while offset < len(data):
        length = struct.unpack('>I', data[offset:offset+4])[0]
        tag = data[offset+4:offset+8]
        body = data[offset+8:offset+8+length]
        if tag == b'IHDR':
            width, height, bitdepth, colortype, compression, filter_method, interlace = struct.unpack('>IIBBBBB', body)
        elif tag == b'IDAT':
            idat_chunks.append(body)
        offset += 12 + length
        
    raw = zlib.decompress(b''.join(idat_chunks))
    
    bpp = 4 # RGBA
    stride = width * bpp
    pixels = bytearray(height * stride)
    
    raw_offset = 0
    for y in range(height):
        filter_type = raw[raw_offset]
        raw_offset += 1
        scanline = raw[raw_offset:raw_offset+stride]
        raw_offset += stride
        
        dest_offset = y * stride
        
        for x in range(stride):
            filt = scanline[x]
            left = pixels[dest_offset + x - bpp] if x >= bpp else 0
            up = pixels[dest_offset - stride + x] if y > 0 else 0
            upleft = pixels[dest_offset - stride + x - bpp] if (y > 0 and x >= bpp) else 0
            
            if filter_type == 0:
                val = filt
            elif filter_type == 1: # Sub
                val = (filt + left) & 0xff
            elif filter_type == 2: # Up
                val = (filt + up) & 0xff
            elif filter_type == 3: # Average
                val = (filt + ((left + up) // 2)) & 0xff
            elif filter_type == 4: # Paeth
                p = left + up - upleft
                pa = abs(p - left)
                pb = abs(p - up)
                pc = abs(p - upleft)
                if pa <= pb and pa <= pc:
                    pr = left
                elif pb <= pc:
                    pr = up
                else:
                    pr = upleft
                val = (filt + pr) & 0xff
            else:
                val = filt
                
            pixels[dest_offset + x] = val
            
    return width, height, pixels

def encode_png(width, height, pixels, filepath):
    def chunk(tag, body):
        return struct.pack('>I', len(body)) + tag + body + struct.pack('>I', zlib.crc32(tag + body) & 0xffffffff)
        
    ihdr = struct.pack('>IIBBBBB', width, height, 8, 6, 0, 0, 0)
    raw_data = bytearray()
    
    stride = width * 4
    for y in range(height):
        raw_data.append(0) # Filter type 0 None
        raw_data.extend(pixels[y*stride:(y+1)*stride])
        
    idat = zlib.compress(bytes(raw_data))
    png = b'\x89PNG\r\n\x1a\n' + chunk(b'IHDR', ihdr) + chunk(b'IDAT', idat) + chunk(b'IEND', b'')
    
    os.makedirs(os.path.dirname(filepath), exist_ok=True)
    with open(filepath, 'wb') as f:
        f.write(png)

def recolor_medieval_clean():
    src_png = r"d:\codingan\A-project-corazon\kamish-dagger-1-20-1-e1560\assets\minecraft\optifine\cit\kamish_dagger\kamish_dagger.png"
    dst_mafia = r"d:\codingan\A-project-corazon\src\main\resources\assets\corazonmod\textures\item\mafia_dagger.png"
    dst_kamish = r"d:\codingan\A-project-corazon\src\main\resources\assets\corazonmod\textures\item\kamish_dagger.png"
    
    width, height, pixels = decode_png(src_png)
    
    for i in range(0, len(pixels), 4):
        r, g, b, a = pixels[i], pixels[i+1], pixels[i+2], pixels[i+3]
        if a < 10:
            continue
            
        lum = (0.299 * r + 0.587 * g + 0.114 * b) / 255.0
        
        is_red_purple = (r > 120 and b > 100) or (r > 150 and g < 100)
        is_yellow_gold = (r > 140 and g > 120 and b < 100)
        
        if is_red_purple:
            if lum > 0.6:
                nr, ng, nb = int(min(255, 255 * lum)), int(min(255, 215 * lum)), int(min(255, 90 * lum))
            else:
                nr, ng, nb = int(min(255, 200 * lum + 30)), int(min(255, 40 * lum)), int(min(255, 50 * lum))
        elif is_yellow_gold:
            nr = int(min(255, lum * 230 + 30))
            ng = int(min(255, lum * 170 + 20))
            nb = int(min(255, lum * 60 + 10))
        else:
            if lum > 0.7:
                nr = int(min(255, lum * 240 + 15))
                ng = int(min(255, lum * 245 + 10))
                nb = int(min(255, lum * 255))
            elif lum > 0.35:
                nr = int(min(255, lum * 170 + 20))
                ng = int(min(255, lum * 180 + 20))
                nb = int(min(255, lum * 200 + 25))
            else:
                nr = int(min(255, lum * 100 + 15))
                ng = int(min(255, lum * 85 + 10))
                nb = int(min(255, lum * 75 + 10))
                
        pixels[i] = nr
        pixels[i+1] = ng
        pixels[i+2] = nb
        
    encode_png(width, height, pixels, dst_mafia)
    encode_png(width, height, pixels, dst_kamish)
    print("[SUCCESS] Clean PNG un-filtered recolor completed!")

if __name__ == '__main__':
    recolor_medieval_clean()
